### PR TITLE
Chart tooltip

### PR DESCRIPTION
**Pull request by dinaNW**
https://projects.invisionapp.com/share/8JSLFQTUM4N#/screens/373112548_Expanded_Graphs_-_Graph_Colors

### CODE REVIEW

File: src/_lib/modules/charts/tooltip/tooltip.directive.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import {Directive, ElementRef, Input, OnChanges, OnInit, Renderer2, SimpleChanges} from '@angular/core';
+
+@Directive({
+    selector: '[nwChartTooltip]'
+})
+export class TooltipDirective implements OnChanges, OnInit {
```
Comment (garethdn) [inline src/_lib/modules/charts/tooltip/tooltip.directive.ts]: Given that we may have a generic tooltip directive in the future I'd recommend calling the class `ChartTooltipDirective` - this also matches the selector naming convention

File: src/_lib/modules/charts/tooltip/tooltip.directive.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    private _width: number;
+    private _height: number;
+
+    constructor(
+        private _elRef: ElementRef,
+        private _renderer: Renderer2) {
+    }
+
+    ngOnInit() {
+        this._width = this._elRef.nativeElement.getBoundingClientRect().width;
```
Comment (garethdn) [inline src/_lib/modules/charts/tooltip/tooltip.directive.ts]: Should only call `getBoundingClientRect()` once and use that to get the width and height

Comment (dinaNW) [reply src/_lib/modules/charts/tooltip/tooltip.directive.ts]: @garethdn Isn't putting it inside ngOnInit() already constricting it to only being called once in the component's lifecycle?

**Comment by dinaNW** [comment]
@garethdn comments addressed, but I'm not sure this:
> When I hover on an element (e.g. circle or line) the tooltip flashed
is correct, i.e the flash seems unrelated to hovering over other elements, and I can't pinpoint what's causing it.

File: src/_lib/modules/charts/tooltip/tooltip.directive.spec.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { ChartTooltipDirective } from './chart-tooltip.directive';
```
Comment (garethdn) [inline src/_lib/modules/charts/tooltip/tooltip.directive.spec.ts]: This filename for this spec should update based on the update to the directive filename, and these imports will need to change. I'd say some useful tests could be written here too.

File: src/app/charts/charts.component.html [hunk 4]
```diff
@@ -141,6 +141,19 @@ <h2 class="page-header">Charts</h2>
...
             </svg>
 
+            <div *ngIf="isHovering && showTooltip.value" nwChartTooltip
+                 class="tooltip in" role="tooltip"
+                 [position]="hoverPosition"
+                 [chartWidth]="chart.width"
+                 [chartHeight]="chart.height"
+                 [chartMargins]="chartMargins">
+                <div class="tooltip-inner">
+                    <h3>Whoa! Awesome...</h3>
```
Comment (garethdn) [inline src/app/charts/charts.component.html]: I'd say to update this to something like "Tooltip title"

File: src/_lib/modules/charts/chart.component.ts [hunk 5]
```diff
@@ -46,7 +46,7 @@ export class ChartComponent implements OnInit, AfterViewInit {
     @ViewChild('hoverOverlay') hoverOverlay: ElementRef;
 
     @Output() nwMousemove: EventEmitter<[number, number]> = new EventEmitter();
-    @Output() nwMouseout: EventEmitter<null> = new EventEmitter();
+    @Output() nwMouseleave: EventEmitter<null> = new EventEmitter();
```
Comment (garethdn) [inline src/_lib/modules/charts/chart.component.ts]: I'm ok with changing the name of this - but you'll need to update all references to this output in Spike. This change is a breaking change so will also require a new major version

File: src/app/charts/charts.component.html (hunk 4 above)
Comment (garethdn) [reply src/app/charts/charts.component.html]: Tooltip title update still pending

**Pull request merged**